### PR TITLE
Add VQQ extractor and JSInterp improvements

### LINKED ISSUE #1666: v.qq.com
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2021.11.10.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

_No response_

### Example URLs

- Single video: https://v.qq.com/x/cover/mzc002007g7aqqt/t0960an30x7.html
- Single video: https://v.qq.com/x/page/g0369qp6w4k.html

### Description

this site is very popular in china so it should be supported just like youku...

### Verbose log

```shell
Download Log for url: https://v.qq.com/x/cover/mzc002007g7aqqt/t0960an30x7.html

Waiting
[generic] t0960an30x7: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] t0960an30x7: Downloading webpage
[generic] t0960an30x7: Extracting information
ERROR: Unsupported URL: https://v.qq.com/x/cover/mzc002007g7aqqt/t0960an30x7.html
Download Worker Finished.
```

### PR DESCRIPTION
**Pull request by elyse0**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

This PR adds a new extractor for v.qq.com using most of the code from WeTv. It currently supports videos, episodes and series n-n

**JSInterp improvements**
While I was implementing this extractor, I noticed that vqq was using a different checksum algorithm; the code was something like this. The thing is that it relies on Javascript's 32 bit bitwise operations, so my python implementation was giving wrong results. 

So I made the necessary changes to make it work in the Jsinterp. In the end I didn't need to use it, but it seemed to me that these changes could maybe still be useful
```js
function getChecksum(payload) {
  checksum = 0;
  for (i = 0; i < payload.length; i++) {
    checksum = (checksum << 5) - checksum + payload.charCodeAt(i);
    checksum = checksum & checksum;
  }
  return checksum;
}
```

Fixes #1666

### Downloading video
```bash
[debug] Command-line config: ['https://v.qq.com/x/page/q326831cny0.html', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.08.19 [48c88e088] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: d9d7436b4
[debug] Python 3.8.10 (CPython 64bit) - Linux-5.4.0-124-generic-x86_64-with-glibc2.29 (glibc 2.31)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg n5.0.1-8-g11eff7739a-20220712 (setts), ffprobe n5.0.1-8-g11eff7739a-20220712, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] [vqq:video] Extracting URL: https://v.qq.com/x/page/q326831cny0.html
[vqq:video] q326831cny0: Downloading webpage
[vqq:video] q326831cny0: Downloading webpage
[vqq:video] q326831cny0: Downloading webpage
[vqq:video] q326831cny0: Downloading m3u8 information
[vqq:video] q326831cny0: Downloading m3u8 information
[vqq:video] q326831cny0: Downloading m3u8 information
[vqq:video] q326831cny0: Downloading m3u8 information
[vqq:video] q326831cny0: Downloading webpage
[vqq:video] q326831cny0: Downloading m3u8 information
[vqq:video] q326831cny0: Downloading m3u8 information
[vqq:video] q326831cny0: Downloading m3u8 information
[vqq:video] q326831cny0: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] q326831cny0: Downloading 1 format(s): 11
[debug] Invoking hlsnative downloader on "https://apd-vlive.apdcdn.tc.qq.com/defaultts.tc.qq.com/svp_50001/R6O9pkuDbRyImuBw-0SYhtyqK2PweBKQ6EWrf0ONZSwpk4UMQyRM23LNKtBWjKUTU1_yUXaeZolFOS6Yr4Pq1TcxD4VIjSj56Ed2SsfIVNg5RCJi_g5TmQtt5Zf0BvLx3zKH-cNix8TCBo0PpPDZ1ns0Iw7f3B8AHDAd0qKJLUY/szg_9529_50001_0b6baqadqaaaaiap6diyrzqvcbgdhacaaoca.f306110.ts.m3u8?ver=4"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 20
[download] Destination: 我是选手：雷霆裂阵，终极时刻 [q326831cny0].mp4
[download] 100% of 60.50MiB in 00:20 at 2.96MiB/s
[debug] ffprobe command line: ffprobe -hide_banner -show_format -show_streams -print_format json 'file:我是选手：雷霆裂阵，终极时刻 [q326831cny0].mp4'
[FixupM3u8] Fixing MPEG-TS in MP4 container of "我是选手：雷霆裂阵，终极时刻 [q326831cny0].mp4"
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i 'file:我是选手：雷霆裂阵，终极时刻 [q326831cny0].mp4' -map 0 -dn -ignore_unknown -c copy -f mp4 -boasc -movflags +faststart 'file:我是选手：雷霆裂阵，终极时刻 [q326831cny0].temp.mp4'
```

### Downloading series episode
```bash
[debug] Command-line config: ['https://v.qq.com/x/cover/mzc00200p29k31e/s0043cwsgj0.html', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.08.19 [48c88e088] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: d9d7436b4
[debug] Python 3.8.10 (CPython 64bit) - Linux-5.4.0-124-generic-x86_64-with-glibc2.29 (glibc 2.31)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg n5.0.1-8-g11eff7739a-20220712 (setts), ffprobe n5.0.1-8-g11eff7739a-20220712, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] [vqq:episode] Extracting URL: https://v.qq.com/x/cover/mzc00200p29k31e/s0043cwsgj0.html
[vqq:episode] s0043cwsgj0: Downloading webpage
[vqq:episode] s0043cwsgj0: Downloading webpage
[vqq:episode] s0043cwsgj0: Downloading webpage
[vqq:episode] s0043cwsgj0: Downloading m3u8 information
[vqq:episode] s0043cwsgj0: Downloading m3u8 information
[vqq:episode] s0043cwsgj0: Downloading m3u8 information
[vqq:episode] s0043cwsgj0: Downloading m3u8 information
[vqq:episode] s0043cwsgj0: Downloading webpage
[vqq:episode] s0043cwsgj0: Downloading m3u8 information
[vqq:episode] s0043cwsgj0: Downloading m3u8 information
[vqq:episode] s0043cwsgj0: Downloading m3u8 information
[vqq:episode] s0043cwsgj0: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] s0043cwsgj0: Downloading 1 format(s): 11
[debug] Invoking hlsnative downloader on "https://apd-vlive.apdcdn.tc.qq.com/defaultts.tc.qq.com/svp_50112/RwMmKoyzZv4HOKYPA7LxbLR9KZ1O3DsBSnskmyUEQoYV967tGl_Wi1DAGsCE4fTfyVVw3GXfNrr50BdP6_DNMrTsK96mMH-XgPF8YqN6smibKARnt-T1OeZKBrK8HOKMjjdU4VMmWITwLav50ri5IxLr-_fgoSLTU45INLaOB3s/gzc_1000102_0b53muaamaaaoqafq2gexjrmazodazuqaasa.f323013.ts.m3u8?ver=4"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 131
[download] Destination: 第1集：如何快乐吃糖？ [s0043cwsgj0].mp4
[download]  16.7% of ~146.80MiB at    1.46MiB/s ETA 00:53 (frag 21/131)
```
<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by pukkandan** [comment]
Don't import ctypes globally. See #4541

File: yt_dlp/jsinterp.py [hunk 1]
```diff
@@ -280,6 +282,10 @@ def _operator(self, op, left_val, right_expr, expr, local_vars, allow_recursion)
             return right_val
 
         try:
+            # https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators#binary_bitwise_operators
+            if op in ('&', '|', '^', '~', '<<', '>>', '>>', '>>>'):
+                return c_int32(_OPERATORS[op](left_val, right_val)).value
```
Comment (Lesmiscore) [inline yt_dlp/jsinterp.py:287]: I believe this does work without ctypes
```suggestion
                return _OPERATORS[op](left_val, right_val) & 0xffffffff
```

Comment (elyse0) [reply yt_dlp/jsinterp.py]: It does 😮

**Comment by pukkandan** [comment]
pls check I haven't broken anything

**Comment by elyse0** [comment]
Yep, everything works ^-^

I made two last changes to the extractors because tests were failing:
- Update series episodes extraction (the page now divides the episode list in two)
. Improve get_clean_title() (i.e. EP1: Cute Programmer - Watch online - WeTV)

File: yt_dlp/extractor/tencent.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            native_subtitles = self._extract_video_native_subtitles(api_response, subtitle_format)
+
+            formats.extend(fmts)
+            self._merge_subtitles(subs, native_subtitles, target=subtitles)
+
+        self._sort_formats(formats)
+        return formats, subtitles
+
+    def _get_clean_title(self, title):
+        return re.sub(r'\s*[_\-]\s*(?:Watch online|腾讯视频|(?:高清)?1080P在线观看平台).*?$', '', title)
```
Comment (pukkandan) [inline yt_dlp/extractor/tencent.py:122]:
```suggestion
        return re.sub(
            r'\s*[_\-]\s*(?:Watch online|腾讯视频|(?:高清)?1080P在线观看平台).*?$',
            '', title or '').strip() or None
```

Comment (dirkf) [reply yt_dlp/extractor/tencent.py]: `strip_or_none()` ?

Comment (pukkandan) [reply yt_dlp/extractor/tencent.py]: that will return `''` instead of `None` when `title=None`

**Pull request merged**